File: supportlib/src/glmaterial.cpp
/* ***** BEGIN LICENSE BLOCK *****
 * Version: MPL 1.1/GPL 2.0/LGPL 2.1
 *
 * The contents of this file are subject to the Mozilla Public License Version
 * 1.1 (the "License"); you may not use this file except in compliance with
 * the License. You may obtain a copy of the License at
 * http://www.mozilla.org/MPL/
 *
 * Software distributed under the License is distributed on an "AS IS" basis,
 * WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
 * for the specific language governing rights and limitations under the
 * License.
 *
 * The Original Code is the Python Computer Graphics Kit.
 *
 * The Initial Developer of the Original Code is Matthias Baas.
 * Portions created by the Initial Developer are Copyright (C) 2004
 * the Initial Developer. All Rights Reserved.
 *
 * Contributor(s):
 *
 * Alternatively, the contents of this file may be used under the terms of
 * either the GNU General Public License Version 2 or later (the "GPL"), or
 * the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
 * in which case the provisions of the GPL or the LGPL are applicable instead
 * of those above. If you wish to allow use of your version of this file only
 * under the terms of either the GPL or the LGPL, and not to allow others to
 * use your version of this file under the terms of the MPL, indicate your
 * decision by deleting the provisions above and replace them with the notice
 * and other provisions required by the GPL or the LGPL. If you do not delete
 * the provisions above, a recipient may use your version of this file under
 * the terms of any one of the MPL, the GPL or the LGPL.
 *
 * ***** END LICENSE BLOCK ***** */

#include "glmaterial.h"

#include "opengl.h"

namespace support3d {

//////////////////////////////////////////////////////////////
// Texture
//////////////////////////////////////////////////////////////

/**
  Constructor.
 */
GLTexture::GLTexture()
: environment_map(false),
  mipmap(true),
  mode(GL_DECAL),
  mag_filter(GL_LINEAR),
  min_filter(GL_LINEAR),
  wrap_s(GL_REPEAT),
  wrap_t(GL_REPEAT),
  internalformat(GL_RGB),
  texenvcolor(1),
  transform(1),
  imagename(),
  texname(0),
  flags(0)
{
}

/**
  Destructor.
 */
GLTexture::~GLTexture()
{
}

/**
  Return the name of the texture image.

  \return Image name
 */
string GLTexture::getImageName()
{
  return imagename;
}

/**
  Set a new texture image name.

  The texture will be reloaded in the next call to applyGL().

  \param name New texture image name
 */
void GLTexture::setImageName(string name)
{
  imagename = name;
  flags |= IMAGE_NAME;
}

/**
  Do the OpenGL calls to activate the texture.
 */
void GLTexture::applyGL()
{
  if (texname==0)
  {
    allocGL();
  }

  glBindTexture(GL_TEXTURE_2D, texname);

  // Has the image name changed? Then reload
  if (flags & IMAGE_NAME)
  {
    flags &= ~IMAGE_NAME;
    loadTexData();
  }

  glTexEnvi(GL_TEXTURE_ENV, GL_TEXTURE_ENV_MODE, mode);
  GLfloat col[4] = {GLfloat(texenvcolor.x), GLfloat(texenvcolor.y),
                    GLfloat(texenvcolor.z), GLfloat(texenvcolor.w)};
  glTexEnvfv(GL_TEXTURE_ENV, GL_TEXTURE_ENV_COLOR, col);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MAG_FILTER, mag_filter);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_MIN_FILTER, min_filter);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_S, wrap_s);
  glTexParameteri(GL_TEXTURE_2D, GL_TEXTURE_WRAP_T, wrap_t);

  glMatrixMode(GL_TEXTURE);
  double M[16];
  glLoadIdentity();
  transform.toList(M);
  glMultMatrixd(M);
  glMatrixMode(GL_MODELVIEW);

  if (environment_map)
  {
    // Enable environment mapping
    glTexGeni(GL_S, GL_TEXTURE_GEN_MODE, GL_SPHERE_MAP);
    glTexGeni(GL_T, GL_TEXTURE_GEN_MODE, GL_SPHERE_MAP);
    glEnable(GL_TEXTURE_GEN_S);
    glEnable(GL_TEXTURE_GEN_T);
  }
  else
  {
    glDisable(GL_TEXTURE_GEN_S);
    glDisable(GL_TEXTURE_GEN_T);
  }
}

/**
  Allocate the OpenGL texture object.
 */
void GLTexture::allocGL()
{
  GLuint tn[1];

  // Alloc texture name and create texture object...
  glGenTextures(1, tn);
  texname = tn[0];
  glBindTexture(GL_TEXTURE_2D, texname);

  std::cout<<"Texname: "<<texname<<std::endl;
}

void GLTexture::releaseGL()
{
}

/**
  Pass the texture data to OpenGL.

  This method has to be called by the texData() method.

  \param w Width of the texture image (must be a power of 2)
  \param h Height of the texture image (must be a power of 2)
  \param format Format of the data (see format parameter of the glTexImage2D() function)
  \param type Type of the data (see type parameter of the glTexImage2D() function)
  \param data Image data
 */
void GLTexture::texData(int w, int h, int format, int type, char* data)
{
  glPixelStorei(GL_UNPACK_ALIGNMENT, 1);
  if (mipmap)
    gluBuild2DMipmaps(GL_TEXTURE_2D, internalformat, w, h, format, type, data);
  else
    glTexImage2D(GL_TEXTURE_2D, 0, internalformat, w, h, 0, format, type, data);
}


//////////////////////////////////////////////////////////////
// Material
//////////////////////////////////////////////////////////////

/**
  Constructor.
 */
GLMaterial::GLMaterial()
 : Material(),
   ambient(vec4d(0.2,0.2,0.2,1), 0),
   diffuse(vec4d(0.7,0.7,0.7,1), 0),
   specular(vec4d(0,0,0,1), 0),
   shininess(0.0, 0),
   emission(vec4d(0,0,0,1), 0),
   blend_sfactor(-1),
   blend_dfactor(-1),
   textures(),
   vertex_shader(),
   fragment_shader()
{
  addSlot("ambient", ambient);
  addSlot("diffuse", diffuse);
  addSlot("specular", specular);
  addSlot("shininess", shininess);
  addSlot("emission", emission);
}

/**
  Constructor.
 */
GLMaterial::GLMaterial(string aname, double adensity)
 : Material(aname, adensity),
   ambient(vec4d(0.2,0.2,0.2,1), 0),
   diffuse(vec4d(0.7,0.7,0.7,1), 0),
   specular(vec4d(0,0,0,1), 0),
   shininess(0.0, 0),
   emission(vec4d(0,0,0,1), 0),
   blend_sfactor(-1),
   blend_dfactor(-1),
   textures(),
   vertex_shader(),
   fragment_shader()
{
  addSlot("ambient", ambient);
  addSlot("diffuse", diffuse);
  addSlot("specular", specular);
  addSlot("shininess", shininess);
  addSlot("emission", emission);
}

/** Destructor. */
GLMaterial::~GLMaterial()
{
}

/**
  Apply the material using OpenGL commands.

  The material settings are only applied to the front faces.

  \pre The appropriate OpenGL context has been made current
 */
void GLMaterial::applyGL()
{
  GLfloat c[4] = {0,0,0,1};
  GLfloat f;

  const vec4d& ac = ambient.getValue();
  c[0] = GLfloat(ac.x);
  c[1] = GLfloat(ac.y);
  c[2] = GLfloat(ac.z);
  c[3] = GLfloat(ac.w);
  glMaterialfv(GL_FRONT, GL_AMBIENT, c);

  const vec4d& dc = diffuse.getValue();
  c[0] = GLfloat(dc.x);
  c[1] = GLfloat(dc.y);
  c[2] = GLfloat(dc.z);
  c[3] = GLfloat(dc.w);
  glMaterialfv(GL_FRONT, GL_DIFFUSE, c);

  const vec4d& sc = specular.getValue();
  c[0] = GLfloat(sc.x);
  c[1] = GLfloat(sc.y);
  c[2] = GLfloat(sc.z);
  c[3] = GLfloat(sc.w);
  glMaterialfv(GL_FRONT, GL_SPECULAR, c);

  const vec4d& ec = emission.getValue();
  c[0] = GLfloat(ec.x);
  c[1] = GLfloat(ec.y);
  c[2] = GLfloat(ec.z);
  c[3] = GLfloat(ec.w);
  glMaterialfv(GL_FRONT, GL_EMISSION, c);

  f = GLfloat(shininess.getValue());
  glMaterialf(GL_FRONT, GL_SHININESS, f);

  boost::shared_ptr<GLTexture> texture = getTexture(0);
  if (texture.get()!=0)
  {
    texture->applyGL();
    glEnable(GL_TEXTURE_2D);
  }
  else
  {
    glDisable(GL_TEXTURE_2D);
  }

  // Blending?
  if (usesBlending())
  {
    glEnable(GL_BLEND);
    glBlendFunc(blend_sfactor, blend_dfactor);
  }
  else
  {
    glDisable(GL_BLEND);
  }
}

/**
  Return True if the material activates blending.
 */
bool GLMaterial::usesBlending()
{
  return (blend_sfactor!=-1) && (blend_dfactor!=-1);
}

/**
  Return the current size of the texture array.

  \return Number of textures
 */
int GLMaterial::getNumTextures() const
{
  return textures.size();
}

/**
  Set a new size for the texture array.

  \param num New number of textures
 */
void GLMaterial::setNumTextures(int num)
{
  if (num<1)
    num=1;
  textures.resize(num);
}


/**
  Get a stored texture.

  The method returns an empty pointer if the given index \a idx is out
  of range or if there is no texture stored at that position.

  \param idx The index of the texture (can be negative to count from the end of the list)
  \return The texture object
 */
boost::shared_ptr<GLTexture> GLMaterial::getTexture(int idx) const
{
  if (idx<0)
    idx = textures.size()+idx;
  if ((idx<0) || (idx>=int(textures.size())))
    return boost::shared_ptr<GLTexture>();

  return textures[idx];
}

/**
  Set a new texture.

  An EIndexError exception is thrown if the index is out of range.

  \param atexture The new texture object
  \param idx The index of the texture (can be negative to count from the end of the list)
 */
void GLMaterial::setTexture(boost::shared_ptr<GLTexture> atexture, int idx)
{
  if (idx<0)
    idx = textures.size()+idx;
  if ((idx<0) || (idx>=int(textures.size())))
    throw EIndexError();

  textures[idx] = atexture;
}

/**
  Return the current size of the vertex shader array.

  \return Number of vertex shader objects
 */
int GLMaterial::getNumVertexShaders() const
{
  return vertex_shader.size();
}

/**
  Set a new size for the vertex shader array.

  \param num New number of vertex shader objects
 */
void GLMaterial::setNumVertexShaders(int num)
{
  if (num<1)
    num=1;
  vertex_shader.resize(num);
}


/**
  Get a stored vertex shader object.

  The method returns an empty pointer if the given index \a idx is out
  of range or if there is no shader object stored at that position.

  \param idx The index of the shader object (can be negative to count from the end of the list)
  \return The shader object
 */
boost::shared_ptr<GLShader> GLMaterial::getVertexShader(int idx) const
{
  if (idx<0)
    idx = vertex_shader.size()+idx;
  if ((idx<0) || (idx>=int(vertex_shader.size())))
    return boost::shared_ptr<GLShader>();

  return vertex_shader[idx];
}

/**
  Set a new vertex shader object.

  An EValueError exception is thrown if \a ashader is not of type VERTEX.
  An EIndexError exception is thrown if the index is out of range.

  \param ashader The new shader object
  \param idx The index of the shader object (can be negative to count from the end of the list)
 */
void GLMaterial::setVertexShader(boost::shared_ptr<GLShader> ashader, int idx)
{
  // Check if the shader is a VERTEX shader...
  if (ashader.get()!=0)
  {
    if (ashader->getType()!=GLShader::VERTEX)
      throw EValueError("Shader must be of type VERTEX");
  }

  if (idx<0)
    idx = vertex_shader.size()+idx;
  if ((idx<0) || (idx>=int(vertex_shader.size())))
    throw EIndexError();

  vertex_shader[idx] = ashader;
}

/**
  Return the current size of the fragment shader array.

  \return Number of fragment shader objects
 */
int GLMaterial::getNumFragmentShaders() const
{
  return fragment_shader.size();
}

/**
  Set a new size for the fragment shader array.

  \param num New number of fragment shader objects
 */
void GLMaterial::setNumFragmentShaders(int num)
{
  if (num<1)
    num=1;
  fragment_shader.resize(num);
}


/**
  Get a stored fragment shader object.

  The method returns an empty pointer if the given index \a idx is out
  of range or if there is no shader object stored at that position.

  \param idx The index of the shader object (can be negative to count from the end of the list)
  \return The shader object
 */
boost::shared_ptr<GLShader> GLMaterial::getFragmentShader(int idx) const
{
  if (idx<0)
    idx = fragment_shader.size()+idx;
  if ((idx<0) || (idx>=int(fragment_shader.size())))
    return boost::shared_ptr<GLShader>();

  return fragment_shader[idx];
}

/**
  Set a new fragment shader object.

  An EValueError exception is thrown if \a ashader is not of type FRAGMENT.
  An EIndexError exception is thrown if the index is out of range.

  \param ashader The new shader object
  \param idx The index of the shader object (can be negative to count from the end of the list)
 */
void GLMaterial::setFragmentShader(boost::shared_ptr<GLShader> ashader, int idx)
{
  // Check if the shader is a FRAGMENT shader...
  if (ashader.get()!=0)
  {
    if (ashader->getType()!=GLShader::FRAGMENT)
      throw EValueError("Shader must be of type FRAGMENT");
  }

  if (idx<0)
    idx = fragment_shader.size()+idx;
  if ((idx<0) || (idx>=int(fragment_shader.size())))
    throw EIndexError();

  fragment_shader[idx] = ashader;
}


}  // end of namespace

File: cgkit/glmaterial.py
# ***** BEGIN LICENSE BLOCK *****
# Version: MPL 1.1/GPL 2.0/LGPL 2.1
#
# The contents of this file are subject to the Mozilla Public License Version
# 1.1 (the "License"); you may not use this file except in compliance with
# the License. You may obtain a copy of the License at
# http://www.mozilla.org/MPL/
#
# Software distributed under the License is distributed on an "AS IS" basis,
# WITHOUT WARRANTY OF ANY KIND, either express or implied. See the License
# for the specific language governing rights and limitations under the
# License.
#
# The Original Code is the Python Computer Graphics Kit.
#
# The Initial Developer of the Original Code is Matthias Baas.
# Portions created by the Initial Developer are Copyright (C) 2004
# the Initial Developer. All Rights Reserved.
#
# Contributor(s):
#
# Alternatively, the contents of this file may be used under the terms of
# either the GNU General Public License Version 2 or later (the "GPL"), or
# the GNU Lesser General Public License Version 2.1 or later (the "LGPL"),
# in which case the provisions of the GPL or the LGPL are applicable instead
# of those above. If you wish to allow use of your version of this file only
# under the terms of either the GPL or the LGPL, and not to allow others to
# use your version of this file under the terms of the MPL, indicate your
# decision by deleting the provisions above and replace them with the notice
# and other provisions required by the GPL or the LGPL. If you do not delete
# the provisions above, a recipient may use your version of this file under
# the terms of any one of the MPL, the GPL or the LGPL.
#
# ***** END LICENSE BLOCK *****
# $Id: glmaterial.py,v 1.6 2006/04/05 08:17:32 mbaas Exp $

## \file glmaterial.py
## Contains the GLMaterial class.

import _core
from cgtypes import *
from _OpenGL.GL import *
from _OpenGL.GLU import *
import math
import sys, os, os.path
import _Image as Image
import glslangparams
from slots import *

GLSLANG_VERTEX = _core.GLShader.ShaderType.VERTEX
GLSLANG_FRAGMENT = _core.GLShader.ShaderType.FRAGMENT

# GLTexture
class GLTexture(_core.GLTexture):
    def __init__(self,
                 imagename = "",
                 image = None,
                 mode = GL_DECAL,
                 mipmap = True,
                 mag_filter = GL_LINEAR,
                 min_filter = GL_LINEAR,
                 wrap_s = GL_REPEAT,
                 wrap_t = GL_REPEAT,
                 internalformat = GL_RGB,
                 texenvcolor = vec4(1),
                 transform = mat4(1),
                 size = None,
                 environment_map = False):
        """Constructor.
        """
        
        _core.GLTexture.__init__(self)
        self.imagename = imagename
        self.mode = mode
        self.mag_filter = mag_filter
        if mipmap and min_filter==GL_LINEAR:
            min_filter = GL_LINEAR_MIPMAP_LINEAR
        self.min_filter = min_filter
        self.wrap_s = wrap_s
        self.wrap_t = wrap_t
        self.texenvcolor = vec4(texenvcolor)
        self.transform = transform
        self.environment_map = environment_map
        self.mipmap = mipmap
        self.internalformat = internalformat
        self.size = size

        self._image = image
        
        # Remember the current working directory
        self._path = os.getcwd()

    # loadTexData
    def loadTexData(self):
        """Load and apply texture file.
        """
        # No image data set? Then use the file name and load the image...
        if self.image==None:
            if self.imagename=="":
                return
            print 'Loading "%s"...'%self.imagename,
            fullname = os.path.join(self._path, self.imagename)
            try:
                img = Image.open(fullname)
                self._passPILImage(img)
            except IOError, e:
                print "failed"
                print e
                return
        else:
            data = self.image
            if type(data)==str:
                # Raw image data as string
                w,h = self.size
                format = self.internalformat
                self.texData(w, h, format, GL_UNSIGNED_BYTE, data)
            else:
                # PIL image
                self._passPILImage(data)


    ## protected:

    def _passPILImage(self, img):
        """Pass a PIL image back to OpenGL.

        img must be a PIL image.
        The image is scaled to a power of 2 if necessary.
        """
        img = self._fitPILImage(img)
        
        w,h = img.size
        
        if img.mode=="RGB":
            format = GL_RGB
        elif img.mode=="RGBA":
            format = GL_RGBA
        else:
            raise RuntimeError, "Unsupported texture format (%s)"%img.mode
        
        self.texData(w, h, format, GL_UNSIGNED_BYTE, img.tostring())
        

    def _fitPILImage(self, img):
        """Scale the image to a power of 2 resolution.

        The image is scale to the user specified resolution. If no
        resolution was set, the next higher power of 2 resolution is used.
        
        If the image resolution is already a power of 2 then
        the input image is returned.
        """
        w,h = img.size
        if self.size==None:
            w2 = 2**int(math.ceil(math.log(w)/math.log(2)))
            h2 = 2**int(math.ceil(math.log(h)/math.log(2)))
        else:
            w2,h2 = self.size
            
        if w2!=w or h2!=h:
            img = img.resize((w2,h2), Image.BICUBIC)

        return img
        
        
    # "image" property...
    
    def _getImage(self):
        """Return the current image.

        This method is used for retrieving the \a image property.

        \return PIL Image (or None)
        """
        return self._image

    def _setImage(self, img):
        """Set a new texture image.

        This method is used for setting the \a image property.

        \param image PIL image or None
        """
        self._image = img
        # Assign the name again so that the image data will be reloaded
        self.imagename = self.imagename

    image = property(_getImage, _setImage, None, "Texture image")

# GLShader
class GLShader(_core.GLShader):
    """OpenGL 2 shader source file.
    
    """

    def __init__(self,
                 shadertype,
                 filename,
                 cpp = None,
                 cpperrstream = sys.stderr,
                 **shaderparams):
        """Constructor.
        """
        _core.GLShader.__init__(self, shadertype, filename)

        # Extract the shader parameters...
        uniform,attribute,varying = glslangparams.glslangparams(filename, cpp=cpp, cpperrstream=cpperrstream)

        self._uniform = {}
        for type,name,arraysize,structname,struct in uniform:
            if type=="struct":
                print "Warning: structs are not yet supported"
                continue
            self._uniform[name] = (type,name,arraysize,structname,struct)

        for paramname in shaderparams:
            setattr(self, paramname, shaderparams[paramname])

    def __str__(self):
        if self.type==GLSLANG_VERTEX:
            stype = "vertex"
        else:
            stype = "fragment"
        return "<%s shader '%s'>"%(stype, os.path.basename(self.filename))

    # getattr
    def __getattr__(self, attr):
        """Return slot value or None if the slot wasn't created yet.

        An attempt an unknwon attribute (i.e. no uniform var) results
        in an error.
        """
        slot = self.__dict__.get("%s_slot"%attr, None)
        if slot!=None:
            return slot.getValue()

        if attr in self._uniform:
            return None

        raise AttributeError, "shader '%s' has no parameter '%s'"%(self.filename, attr)

    # setattr
    def __setattr__(self, attr, value):
        if attr in ["_uniform"] or attr[-5:]=="_slot":
            _core.GLShader.__setattr__(self, attr, value)
            return

        # Is attr a slot variable?
        slot = self.__dict__.get("%s_slot"%attr, None)
        if slot!=None:
            slot.setValue(value)
            return 

        # Is it a parameter that has no value yet? then we must create
        # a new slot...
        uniform = getattr(self, "_uniform", [])
        if attr in uniform:
            slot = self._createSlot(attr, uniform[attr][0])
            slot.setValue(value)
        else:
            raise AttributeError, "shader '%s' has no parameter '%s'"%(self.filename, attr)

    # iterUniform
    def iterUniform(self):
        return self._uniform.itervalues()

    # _createSlot
    def _createSlot(self, param, type):
        """Create a slot for a parameter.

        param is the name of the parameter and type its OpenGL type (as str).
        """
        slot_lut = {"bool" : "BoolSlot",
                    "float" : "DoubleSlot",
                    "int" : "IntSlot",
                    "bvec2" : "Vec3Slot",
                    "bvec3" : "Vec3Slot",
                    "bvec4" : "Vec4Slot",
                    "ivec2" : "Vec3Slot",
                    "ivec3" : "Vec3Slot",
                    "bvec4" : "Vec4Slot",
                    "vec2" : "Vec3Slot",
                    "vec3" : "Vec3Slot",
                    "vec4" : "Vec4Slot",
                    "mat2" : "Mat3Slot",
                    "mat3" : "Mat3Slot",
                    "mat4" : "Mat4Slot",
                    "sampler1D" : "IntSlot",
                    "sampler2D" : "IntSlot",
                    "sampler3D" : "IntSlot",
                    "samplerCube" : "IntSlot",
                    "sampler1DShadow" : "IntSlot",
                    "sampler2DShadow" : "IntSlot"}
        slotname = slot_lut.get(type, None)
        if slotname==None:
            raise ValueError, "Invalid parameter type: %s"%type
        exec "slot = %s()"%slotname
        self.addSlot(param, slot)
        # Store the slot as <param>_slot 
        setattr(self, "%s_slot"%param, slot)
        return slot

        
# GLMaterial
class GLMaterial(_core.GLMaterial):

    def __init__(self,
                 name = "GLMaterial",
                 ambient = (0.2, 0.2, 0.2, 1),
                 diffuse = (0.7, 0.7, 0.7, 1),
                 specular = (0,0,0,1),
                 shininess = 0.0,
                 emission = (0,0,0,1),
                 blend_factors = None,
                 texture = None,
                 vertex_shader = None,
                 fragment_shader = None,
                 density=1.0):
        _core.GLMaterial.__init__(self, name, density)

        self.ambient = vec4(ambient)
        self.diffuse = vec4(diffuse)
        self.specular = vec4(specular)
        self.shininess = shininess
        self.emission = vec4(emission)
        if texture!=None:
            if isinstance(texture, _core.GLTexture):
                texture = [texture]
            self.setNumTextures(len(texture))
            for i,tex in enumerate(texture):
                self.setTexture(tex,i)
        if vertex_shader!=None:
            if isinstance(vertex_shader, _core.GLShader):
                vertex_shader = [vertex_shader]
            self.setNumVertexShaders(len(vertex_shader))
            for i,shd in enumerate(vertex_shader):
                self.setVertexShader(shd,i)
        if fragment_shader!=None:
            if isinstance(fragment_shader, _core.GLShader):
                fragment_shader = [fragment_shader]
            self.setNumFragmentShaders(len(fragment_shader))
            for i,shd in enumerate(fragment_shader):
                self.setFragmentShader(shd,i)
        if blend_factors!=None:
            self.blend_sfactor = blend_factors[0]
            self.blend_dfactor = blend_factors[1]
